README.md
# Payshares Data API
The Payshares data API is the end point for payshares-charts-frontend and other applications that need historical data.  This API is built on Node.js, CouchDB, and Redis.

## Components

### Ledger Importer
The ledger importer imports ledgers from the Payshares Network into the data store.  The process is set up to import continously in real time as ledgers are validated, as well as import historical ledgers.

### Data Store
The data store uses CouchDB to store ledgers as a JSON document.  To retrieve structured data from the ledgers, "views" have been created which index specific elements from the ledger into a format that can be queried efficiently.  For example, one view collects all successful transactions that exercise offers and indexes them according to the currencies, issuers, and date, and returns the amounts and prices of the exercised offers.  We can then query this view and group the indexed results by time increments to get an array of offers excercised over time for a given pair of currencies.

### API Server
Accessing the historical data is not done by querying the database directly but through a node.js API server.  The server takes requests in the form of JSON data, interprets it into one or several couchDB queries, processes the data as necessary and returns the results.

### Memory Cache
To reduce the load on CouchDB, the API server contains a cacheing layer.  The cache expects the data stored in couch to be accurate and up to date, thus it will be automatically disabled if the importer is unable to keep up to date for any reason, and will restart again when the data gets caught up.

## Setup
  + install node.js
  + install couchDB or get hosted couchDB service and create new a database 
  + install redis (optional)
```  
    git clone https://github.com/payshares/payshares-data-api.git
    cd payshares-data-api
    npm install
    cp db.config.json.example db.config.json
    cp deployment.environments.json.example deployment.environments.json
```
  + update the config files to point to your couchdb and redis
  + `grunt updateViews` push the design documents to couchDB, write permissions required 
  + `grunt` or `grunt watch` starts the API service   
  + `node db/importer live debug2` starts the importer service, write permissions required
  + `node db/import` starts the experimental websocket based importer, write permissions required

## group API Routes
All API routes are post requests with parameters passed as a JSON object.  the specific endpoint is defined as `/[ROUTE]`, such as `/offersExercised`

### offers_exercised [/offers_exercised{base}{counter}{startTime}{endTime}{timeIncrement}{timeMultiple}{descending}{reduce}{limit}{offset}{format}]
Exchange offers exercised over time - returns volume in terms of base and counter currencies, number of trades, open, high, low, close, and volume weighted price.

+ Maximum returned results : 500

#### POST    

+ Parameters
    + base (JSON) ... base currency of the trading pair, in the form {currency,issuer}
    + counter (JSON) ... counter currency of the trading pair, in the form {currency, issuer}
    + startTime (string) ... any momentjs-readable date
    + endTime (string) ... any momentjs-readable date
    + timeIncrement (string, optional) ... any of the following: "all", "none", "year", "month", "day", "hour", "minute", "second" - defaults to "day"
    + timeMultiple (integer, optional) ... 5, 15, etc.. defaults to 1 (for 5 minutes, 15 mintues, etc.)
    + descending (boolean, optional) ... defaults to false
    + reduce (boolean, optional) ... defaults to true, false returns individual transactions
    + limit (integer, optional) ... ignored unless reduce is false - limit the number of returned trades
    + offset (integer, optional) ... offset by n transactions for pagination, ignored if reduce != false
    + format ('json' or 'csv', optional) ... default will be a lightweight CSV like array

+ Request (json)

        {    
            base : {currency: "XPS"},
            counter : {currency: "USD", issuer: "gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q"},
            startTime : "2014-03-11",
            endTime : "2014-03-12",
            timeIncrement : "minute",
            timeMultiple : 15,
            format : "json"
        }

+ Response 200 (text/plain) 

    + startTime - start time of period
    + baseVolume - volume in terms of base currency
    + counterVolume - volume in terms of trade currency
    + count - number of trades
    + open - price of the first trade in the time period
    + high - highest price
    + low - lowest price    
    + close - price of the last trade in the time period
    + vwap - volume weighted average price
    + openTime - time of the first trade
    + closeTime - time of the last trade
    + partial - true indicates this row may not include all trades in the interval due to the alignment of the requested time period. For example, a time interval of 1 minute was requested, but the end time for the range was 3:45:30    

    
    + Body
    
            [
                [
                    "startTime",
                    "baseVolume",
                    "counterVolume",
                    "count",
                    "open",
                    "high",
                    "low",
                    "close",
                    "vwap",
                    "openTime",
                    "closeTime",
                    "partial"
                ],
                [
                    "2014-03-11T12:00:00+00:00",
                    283100.785902,
                    4148.880165613894,
                    129,0.014637002338207711,
                    0.01473544628685632,
                    0.01451000001204874,
                    0.01455619191687749,
                    0.014655231083925739,
                    "2014-03-11T12:09:10+00:00",
                    "2014-03-11T15:57:20+00:00",
                    false
                ],
                .
                .
                .
            ]   
            
+ Response 200 (reduce = false, text/plain)

        [
            [ "time", "price", "baseAmount", "counterAmount", "tx_hash" ],
            [ "2014-03-14T20:58:45+00:00", 0.0146372844753653, 346.697779, 5.07471400683615, KAXs2Adoas3..... ],
            .
            .
            .
        ]
        
+ Response 200 (json) 

        {
            startTime : "2014-03-11T11:44:00+00:00",
            endTime   : "2014-03-12T12:09:00+00:00",
            base      : {
                currency : "XPS"
            },
            counter : {
                currency : "USD",
                issuer   : "gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q"
            },
            timeIncrement : "hours",
            timeMultiple  : 4,
            results       : [
                {
                    startTime     : "2014-03-11T12:00:00+00:00",
                    openTime      : "2014-03-11T15:57:20+00:00",
                    closeTime     : "2014-03-11T12:09:10+00:00",
                    baseVolume    : 283100.785902,
                    counterVolume : 4148.880165613894,
                    count         : 129,
                    open          : 0.014637002338207711,
                    high          : 0.01451000001204874,
                    low           : 0.01455619191687749, 
                    close         : 0.01473544628685632,
                    vwap          : 0.014655231083925739,
                    partial       : false
                },
                .
                .
                .
            ]
        }
        
+ Response 200 (text/csv)   

        startTime, baseVolume, counterVolume, count, open, high, low, close, vwap, openTime, closeTime, partial
        2014-03-11T08:00:00+00:00, 457.81426600000003, 6.701129377783959, 4, 0.0146372844753653, 0.0146372844753653, 0.01463700234192037, 0.01463700234192037, 0.014637222778328018, 2014-03-11T11:44:10+00:00, 2014-03-11T11:57:20+00:00, true
        2014-03-11T12:00:00+00:00, 283100.785902, 4148.880165613894, 129, 0.014637002338207711, 0.01473544628685632, 0.01451000001204874, 0.01455619191687749, 0.014655231083925737, 2014-03-11T12:09:10+00:00, 2014-03-11T15:57:20+00:00, false
        2014-03-11T16:00:00+00:00, 265984.147741, 3848.099297442456, 66, 0.01455, 0.01472737999181529, 0.014314547875005369, 0.014580006546125387, 0.014467756083227484, 2014-03-11T16:23:20+00:00, 2014-03-11T19:30:10+00:00, false
        ...
        ...
        ...

### value_sent [/value_sent{currency}{issuer}{startTime}{endTime}{timeIncrement}{descending}{reduce}{limit}{offset}{format}]
The amount of value sent from all accounts for a specific currency over time. 

#### POST

+ Parameters
    + currency (string) ... "XPS", "USD", etc.
    + issuer   (string) ... "rvYAfWj5gh67oV6f..." 
    + startTime (string) ... any momentjs-readable date
    + endTime (string) ... any momentjs-readable date
    + timeIncrement (string, optional) ... any of the following: "all", "none", "year", "month", "day", "hour", "minute", "second" - defaults to "day"
    + descending (boolean, optional) ... defaults to false
    + reduce (boolean, optional) ... defaults to true, ignored if timeIncrement is set
    + limit (integer, optional) ... limit the number of responses, ignored if reduce != false
    + offset (integer, optional) ... offset by n transactions for pagination, ignored if reduce != false
    + format ('json' or 'csv', optional) ... default will be a lightweight CSV like array

+ Request (json)
        
        {
            currency  : "USD",
            issuer    : "gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q",
            startTime : "Mar 5, 2014 10:00 am",
            endTime   : "Mar 6, 2014 10:00 am"
        }


+ Response 200 (text/plain)

        [
            ["time","amount","count"],
            ["2014-03-04T08:00:00+00:00",1444.360203839022,4],
            ["2014-03-04T09:00:00+00:00",41678.45554351887,12],
            ["2014-03-04T10:00:00+00:00",4095.8676436164487,19],
            ...
            ...
            ...
        ]
        
+ Response 200 (reduce = false)

        [
            [
                "time",
                "amount",
                "account",
                "destination",
                "txHash",
                "ledgerIndex"
            ],
            [
                "2014-03-15T17:19:00+00:00",
                0.151374648,
                "rHsZHqa5oMQNL5....",
                null,
                "BFED0E5164F4....",
                5536660
            ],  
            .
            .
            .
        ]
        
+ Response 200 (json)

        {
            currency      : "BTC",
            issuer        : "gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q",
            startTime     : "2014-03-05T18:00:00+00:00",
            endTime       : "2014-03-06T18:00:00+00:00",
            timeIncrement : "hour",
            results       : [
                {
                    time   : "2014-03-05T18:00:00+00:00",
                    amount : 0.01065677430038825,
                    count  : 1
                },
                {
                    time   : "2014-03-05T19:00:00+00:00",
                    amount : 0.3742335806999557,
                    count  : 3
                },
                .
                .
                .
            ]
        }

+ Response 200 (csv)

        time, amount, count
        2014-03-15T00:00:00+00:00, 296151.704209, 57
        2014-03-15T01:00:00+00:00, 18227.868377, 57
        2014-03-15T02:00:00+00:00, 471575.79683899996, 78
        ...
        ...
        ...

### issuer_capitalization [/issuer_capitalization{currencies}{startTime}{endTime}{timeIncrement}]
Returns the total capitalization (outstanding balance) of a specified issuer & specified currency pair, over the given time range.

+ Maximum currencies per query : 10
+ Maximum returned results : 500 per currency

#### POST

+ Parameters
    + currencies (JSON) ... list of currencies in the form {currency, issuer}
    + startTime (string) ... any momentjs-readable date
    + endTime (string) ... any momentjs-readable date
    + timeIncrement (string, optional) ... any of the following: "all", "none", "year", "month", "day", "hour", "minute", "second" - defaults to "day"
    
+ Request (json)

        {
            currencies : [
              {currency:"USD", issuer:"gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q"},
              {currency:"USD", issuer:"rMwjYedjc7qqtKYVLiAccJSmCwih4LnE2q"},
            ],
            
            startTime : "Mar 1, 2014 10:00 am",
            endTime   : "Apr 1, 2014 10:00 am",
            timeIncrement : "day"
        }

+ Response 200 (json)

    + currency - currency of pair from request
    + issuer - issuer of pair from request
    + name - name of gateway requested, if known
    + results - list of balances over time for the selected currency in the form [unixTimestamp,balance]
    
    + Body
    
            [
                {
                    currency : "USD",
                    issuer   : "gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q",
                    name     : "Bitstamp",
                    results : [
                        ["2013-02-01T00:00:00+00:00",0],
                        ["2013-03-01T00:00:00+00:00",25669.508235252037],
                        ["2013-04-01T00:00:00+00:00",58557.238743623835],
                        ...
                        ...
                        ...
                    ]
                },
                ...
                ...
                ...
            ] 

### accounts_created [/accounts_created{startTime}{endTime}{timeIncrement}{descending}{reduce}{limit}{offset}{format}]
The number of payshares accounts that have been created over time.

+ Max returned results : 500

#### POST

+ Parameters
    + startTime (string, optional) ... any momentjs-readable date, defaults to 30 days before end time
    + endTime (string, optional) ... any momentjs-readable date, defaults to now
    + timeIncrement (string, optional) ... any of the following: "all", "none", "year", "month", "day", "hour", "minute", "second" - defaults to "day"
    + descending (boolean, optional) ... defaults to false
    + reduce (boolean, optional) ... defaults to true, false returns individual accounts
    + limit (integer, optional) ... limit the number of responses, ignored if reduce != false
    + offset (integer, optional) ... offset by n transactions for pagination, ignored if reduce != false
    + format ('json' or 'csv', optional) ... default will be a lightweight CSV like array
    
+ Request (json)

        {
            startTime: "2013-04-08",
            endTime: "2014-04-08",
            timeIncrement: "day,
            descending: true
        }
        
+ Response 200 (text/plain)

        [
            ["time","accountsCreated"]
            ["2013-04-08T00:00:00+00:00",38],
            ["2013-04-07T00:00:00+00:00",53],
            ["2013-04-06T00:00:00+00:00",59],
            ...
            ...
            ...
        ]

+ Response 200 (reduce = false, text/plain) 
    
        [
            ["time","account","txHash","ledgerIndex"],
            [
                "2014-03-09T17:07:50+00:00",
                "rNt4o6wkK5dg...",
                "ADFA23A4E361ED...",
                5426006
            ],
            [
                "2014-03-09T17:08:50+00:00",
                "rNtApu2u698Phy...",
                "CA7E143C6B47FAD7....",
                5426016
            ],
            .
            .
            .
        ]

+ Response 200 (json)

        {
            startTime     : "2014-03-09T17:00:00+00:00",
            endTime       : "2014-03-10T17:00:00+00:00",
            timeIncrement : "hour",
            total         : 172,
            results       : [
                {
                    time  : "2014-03-09T17:00:00+00:00",
                    count : 4
                },
                {
                    time  : "2014-03-09T18:00:00+00:00",
                    count : 2
                },
                .
                .
                .
            ]
        }
        
+ Response 200 (csv)

        time, count
        2014-03-09T17:00:00+00:00, 4
        2014-03-09T18:00:00+00:00, 2
        2014-03-09T19:00:00+00:00, 2
        ...
        ...
        ...

### transaction_stats [/transaction_stats{startTime}{endTime}{timeIncrement}{descending}{reduce}{limit}{offset}{format}]
Breakdown of valid transactions by type on the payshares network over time.

#### POST

+ Parameters
    + startTime (string, optional) ... any momentjs-readable date
    + endTime (string, optional) ... any momentjs-readable date
    + timeIncrement (string, optional) ... any of the following: "all", "none", "year", "month", "day", "hour", "minute", "second" - defaults to "day"
    + descending (boolean, optional) ... defaults to false
    + reduce (boolean, optional) ... defaults to true, false returns individual transactions
    + limit (integer, optional) ... limit the number of responses, ignored if reduce != false
    + offset (integer, optional) ... offset by n transactions for pagination, ignored if reduce != false
    + format ('json' or 'csv', optional) ... default will be a lightweight CSV like array
    
+ Request (json)

        {
            startTime     : "Mar 10, 2014 4:35 am",
            endTime       : "Mar 11, 2014 5:10:30 am",
            timeIncrement : "hour",
            format        : "json"
        }
        
+ Response 200 (text/plain)

        {
            ["time", "Payment", "OfferCreate", "OfferCancel", "TrustSet", "AccountSet", "SetFee", "SetRegularKey"]
            [ "2014-02-28T18:00:00+0000", 502, 244, 102, 83, 12, 2, 5 ],
            [ "2014-02-28T17:00:00+0000", 1800, 500, 232, 103, 55, 12, 4 ],
            [ "2014-02-28T16:00:00+0000", 6102, 1293, 503, 230, 100, 14, 5 ],
            ...
            ...
            ...
        }
        
+ Response 200 (reduce = false, text/plain)

        {
            ["time", "type", "account", "txHash", "ledgerIndex"],
            ["2014-02-28T18:00:00+0000", "Payment",     "rXaaFst....", "4ABA3B0777E97BDEA924A732A943B169D...."],
            ["2014-02-28T17:00:00+0000", "OfferCreate", "rXaaFst....", "4ABA3B0777E97BDEA924A732A943B169D...."],
            ...
            ...
            ...
        }

+ Response 200 (json)

        {
            startTime     : "2014-03-10T11:35:00+00:00",
            endTime       : "2014-03-11T12:10:30+00:00",
            timeIncrement : "hour",
            results       : [
                {
                    Payment     : 292,
                    OfferCreate : 315,
                    OfferCancel : 251,
                    AccountSet  : 18,
                    TrustSet    : 6,
                    time        : "2014-03-10T11:00:00+00:00"
                },
                .
                .
                .
            ]
        }
        
+ Response 200 (text/csv)

        time, TrustSet, Payment, OfferCreate, AccountSet, OfferCancel, SetFee, SetRegularKey
        2013-01-01T00:00:00+00:00, 241, 434, 72, 62, 41, 0, 0
        2013-02-01T00:00:00+00:00, 886, 1890, 3086, 9, 2455, 0, 0
        2013-03-01T00:00:00+00:00, 1537, 3968, 11244, 5, 9390, 0, 0
        ...
        ...
        ...

### exchange_rates [/exchange_rates{pairs}{base}{counter}{range}{last}]
The exchange rates between two currencies for a given time period.  Returns the volume weighted average price for the specified range, as well as the last traded price.

+ Max returned results : 20

#### POST  

+ Parameters 
    + pairs (JSON, optional) ... Array of currency pairs for retreival.  Required if base and counter are absent
    + base (JSON, optional) ... base currency of the trading pair, in the form {currency,issuer}.  Required if `pairs` is absent
    + counter (JSON, optional) ... counter currency of the trading pair, in the form {currency, issuer}.  Required if `pairs` is absent
    + range (string, optional) ... "hour", "day", "week", "month", "year".  Time range to average the price over, defaults to `day`
    + last (boolean, optional) ... true returns the last price only (faster response time)
 
+ Request (json)
    
        {
            pairs : [
                {
                    base    : {currency:"CNY","issuer":"rnuF96W4SZoCJmbHYBFoJZpR8eCaxNvekK"},
                    counter : {currency:"XPS"}
                }
            ],
            range : "day"
        }


+ Response 200 (json)

    + Pairs - array of results as json for each trading pair requested
        + rate  - volume weighted average price for the given range
        + last  - last traded price for the given range
    
    + Body       
    
            {
                pairs : [
                  {
                    base    : {
                        currency : "CNY", 
                        issuer   : "rnuF96W4SZoCJmbHYBFoJZpR8eCaxNvekK",
                        name     : "paysharesCN"
                    },
                    counter : {
                        currency : "XPS"
                    },
                    rate    : .5412
                    last    : .5405
                    range   : "day"
                  },
                    ....
                ] 
            }

### market_traders [/market_traders{base}{counter}{range}{startTime}{format}]
Returns a list of accounts that participated in trading the specified trading pair during the specified time range, ordered by base currency volume.  If no trading pair is provided, the API uses a list of the top XPS markets

#### POST

+ Parameters
    + base (JSON, optional) ... base currency-issuer. If not present, top XPS markets are queried
    + counter  (JSON, optional) ... counter currency-issuer. Required if base is present
    + range (string, optional) ... Any of the following ("24hr", "3d", "7d")
    + startTime (string, optional) ... moment.js readable date string
    + format ('json' or 'csv', optional) ... defaults to a CSV-like array
    
+ Request (json)

        {
            base : {
                currency : "XPS"
            },
            counter : {
                currency : "USD",
                issuer   : "gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q"
            },
            range : "24h"
        }
        
+ Response 200 (text/plain)

        [
            ["account","volume","count"],
            ["rnErNnnxRxpki...",1928.618644523547,8],
            ["rLqAVKdGpJt2...",1586.0617068572724,5],
            ["rGEDQD48uACC2...",1318,3],
            ...
            ...
            ...
        ]
        
+ Response 200 (json)

        {
            startTime : "2014-04-09T20:53:30+00:00",
            endTime   : "2014-04-10T20:53:30+00:00",
            results   : [
                {
                    account : "rwLezhH....",
                    volume  : 9058276.664807992,
                    count   : 219
                },
                {
                    account : "rHrSaVx....",
                    volume  : 4347587.137745989,
                    count   : 456
                },
                .
                .
                .
            ]    
        }

+ Response 200 (csv)

        account, volume, count
        rwZewhQae..., 9484699.918622022, 245
        rHrSaQVSj..., 2717824.9351769933, 245
        rCvBeHsW3..., 2242708.7111989968, 329
        ...
        ...
        ...
        
### total_network_value [/total_network_value{time}{exchange}]
Total value of currencies for the top gateways on the payshares network, normalized to a specified currrency.

#### POST

+ Parameters
    + time (string) ... any momentjs-readable date.  The time of desired snapshot
    + exchange (JSON, optional) ... desired currency for valuation, in the form {currency, issuer}. Defaults to XPS

+ Request (json)

        {
            time      : "2014-03-13T20:39:26+00:00"
            exchange  : {
                currency  : "CNY",         
                issuer    : "rnuF96W4SZoCJmbHYBFoJZpR8eCaxNvekK"
            }
        }

+ Response 200 (json)

    + time - the time of desired snapshot
    + exchange - valuation currency and issuer
    + exchangeRate - final valuation exchange rate
    + total - combined valuation amount in specified valuation currency
    + components - individual component currencies that make up the combined total
        
    + Body   
    
            {
                time     : "2014-03-13T20:39:26+00:00",
                exchange : {
                    currency : "USD", 
                    issuer   : "gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q"
                },
                exchangeRate : 0.014301217579817786,
                total        : 726824.6504823748,
                components   : [
                    {
                        currency        : "USD",
                        issuer          : "gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q",
                        amount          : 27606.296227064257,
                        rate            : 1,
                        convertedAmount : 27606.296227064257
                    },
                    .
                    .
                    .
                ]
            }

### top_markets [/top_markets{startTime}{endTime}{exchange}]
The total trading volume for the top 5 markets on the payshares network for a given time period, normalized USD.  Returns data for the last 24 hours if no arguments are given.

#### POST  

+ Parameters 
    + startTime (string, optional) ... any momentjs-readable date, defaults to 1 day before end time
    + endTime (string, optional) ... any momentjs-readable date, defaults to now
    + exchange (JSON, optional) ... desired currency for valuation, in the form {currency, issuer}. Defaults to XPS
 
+ Request (json)

        {
            startTime : "2014-01-15 7:00 AM",
            endTime   : "2014-01-16 8:00 PM",
            exchange  : {
                currency : "USD",
                issuer : "gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q"
            }
        }

+ Response 200 (json)
    
        { 
            startTime    : '2014-03-13T20:26:24+00:00',
            endTime      : '2014-03-14T20:26:24+00:00',
            exchange     : { currency: 'XPS' },
            exchangeRate : 1,                 
            total        : 1431068.4284775178,
            count        : 627,
            components   : [
                { 
                    base : {
                        currency:"USD",
                        issuer:"gDSSa75HPagWcvQmwH7D51dT5DPmvsKL4q"
                    },
                    counter : {
                        currency:"XPS"
                    },
                    rate            : 69.9309953931345,
                    count           : 99,
                    amount          : 3107.9273091242917,
                    convertedAmount : 217340.45033656774 
                },
                .
                .
                .
            ]
        } 

### total_value_sent [/total_value_sent{startTime}{endTime}{exchange}]
The total of amounts sent or exchanged from any wallet, either through a payment or an "offerCreate" that exercises another offer, for a curated list of currency/issuers and XPS, normalized to a specified currency

#### POST  

+ Parameters 
    + startTime (string, optional) ... any momentjs-readable date, defaults to 1 day before end time
    + endTime (string, optional) ... any momentjs-readable date, defaults to now

release-notes.md
# Release Notes

## Development

[Commits](https://github.com/payshares/payshares-data-api/commits/master)

##December 2014

###New Features and Changes

 - [2](https://github.com/payshares/payshares-data-api/pull/2) Add API for XPS and USD for bitcoin charts  
 - [1](https://github.com/payshares/payshares-data-api/pull/1) Ports and paysharesizes ripple charts for Payshares 
 









api/routes/offers.js
var winston = require('winston'),
  moment    = require('moment'),
  payshares    = require('payshares-lib'),
  _         = require('lodash'),
  tools     = require('../utils');

/*
 * offers:
 * 
 * Returns all offer creates and cancels over time for a given trading pair.
 * 
 * 
   curl -H "Content-Type: application/json" -X POST -d '{
    "base"  : {"currency": "XPS"},
    "counter" : {"currency": "USD", "issuer" : "rvYAfWj5gh67oV6fW32ZzP3Aw4Eubs59B"},
    "startTime" : "Feb 10, 2014 4:44:00 am",
    "endTime"   : "Feb 11, 2014 5:09:00 am",
    "timeIncrement" : "hour",
    "format" : "csv"
    
    }' http://localhost:5993/api/offers
    

   curl -H "Content-Type: application/json" -X POST -d '{
    "base"  : {"currency": "XPS"},
    "counter" : {"currency": "USD", "issuer" : "rvYAfWj5gh67oV6fW32ZzP3Aw4Eubs59B"},
    "startTime" : "Feb 10, 2014 4:44:00 am",
    "endTime"   : "Feb 11, 2014 5:09:00 am",
    "timeIncrement" : "hour",
    "format" : "json"
    
    }' http://localhost:5993/api/offers

   curl -H "Content-Type: application/json" -X POST -d '{
    "base"  : {"currency": "XPS"},
    "counter" : {"currency": "USD", "issuer" : "rvYAfWj5gh67oV6fW32ZzP3Aw4Eubs59B"},
    "startTime" : "Feb 10, 2014 4:44:00 am",
    "endTime"   : "Apr 16, 2014 5:09:00 am",
    "reduce"    : false

     
    }' http://localhost:5993/api/offers 
 * 
 */
function offers (params, callback) {
  
  var options  = {};
  options.view = {};
  parseOptions(); //parse request params for query options
  
  if (options.error) return callback(options.error);  
  
  db.view("offers", "v1", options.view, function (error, couchRes){
    
    if (error) return callback ('CouchDB - ' + error);
    
    handleResponse(couchRes.rows);
  });

  
/*
 * handleResponse - format the data according to the requirements
 * of the request and return it to the caller.
 * 
 */  
  function handleResponse (rows) {  
    
    if (params.format === 'json') {
      var response = {
        base          : params.base,
        counter       : params.counter,
        startTime     : options.startTime.format(),
        endTime       : options.endTime.format(),
        timeIncrement : params.timeIncrement,
        results       : []
      };
      
      if (params.timeIncrement) response.timeIncrement = params.timeIncrement;

      rows.forEach(function(row){
      
        if (options.view.reduce === false) {
          response.results.push({
            type          : row.value[0],
            account       : row.value[1],
            baseAmount    : row.value[2],
            counterAmount : row.value[3],
            price         : row.value[4],
            time          : moment.utc(row.value[5]).format(),
            txHash        : row.value[6],
            ledgerIndex   : parseInt(row.id, 10)
          });
          
        } else {
          response.results.push({
            time : row.key ? moment.utc(row.key.slice(1)) : options.startTime,
            OfferCreate : row.value[0],
            OfferCancel : row.value[1],
          });    
        }
      });     
      
      return callback(null, response);
      
    } else {
      
    
      var header, results = [];
      if (options.view.reduce !== false) header = ["time", "OfferCreate", "OfferCancel"];
      else header = ["type", "account",'baseAmount','counterAmount','price','time','txHash','ledgerIndex'];
   
      console.log(rows);
      results.push(header);
      rows.forEach(function(row){
        var value = row.value;
      
        if (options.view.reduce === false) {
          value[5] = moment.utc(value[5]).format();
          value.push(parseInt(row.id, 10));
          
        } else {
          var time  = row.key ? moment.utc(row.key.slice(1)) : options.startTime;
          value.unshift(time.format());        
        }
  
        results.push(value);
      }); 
      
      if (params.format === 'csv') {

        var csvStr = _.map(results, function(row){
          return row.join(', ');
        }).join('\n');
  
        // provide output as CSV
        return callback(null, csvStr);
      
      } else return callback(null, results);
    }
  }
  
  function parseOptions() {
     if (!params.base || !params.counter) {
      options.error = 'please specify base and counter currencies';
      return;
    }
  
    // parse base currency details
    var base, counter;
    
    if (!params.base.currency) {
        options.error = 'please specify a base currency';
        return;
      
    } else if (!params.base.issuer) {
      
      if (params.base.currency.toUpperCase() === 'XPS') {
        options.base = 'XPS';
      } else {
        options.error = 'must specify issuer for all currencies other than XPS';
        return;
      }
      
    } else if (params.base.issuer && payshares.UInt160.is_valid(params.base.issuer)) {
      options.base = params.base.currency.toUpperCase()+"."+params.base.issuer;
      
    } else {
      var baseGatewayAddress = gatewayNameToAddress(params.base.issuer, params.base.currency.toUpperCase());
      if (baseGatewayAddress) {
        options.base = params.base.currency.toUpperCase()+"."+baseGatewayAddress;
        
      } else {
        options.error = 'invalid base currency issuer: ' + params.base.issuer;
        return;
      }
    }
  
    // parse counter currency details
    if (!params.base.currency) {
      options.error = 'please specify a base currency';
      return;
      
    } else if (!params.counter.issuer) {
      if (params.counter.currency.toUpperCase()  === 'XPS') {
        options.counter = 'XPS';
      } else {
        options.error = 'must specify issuer for all currencies other than XPS';
        return;
      }
    } else if (params.counter.issuer && payshares.UInt160.is_valid(params.counter.issuer)) {
      options.counter = params.counter.currency.toUpperCase()+"."+params.counter.issuer;
      
    } else {
      var counterGatewayAddress = gatewayNameToAddress(params.counter.issuer, params.counter.currency.toUpperCase());
      if (counterGatewayAddress) {
        options.counter = params.counter.currency.toUpperCase()+"."+counterGatewayAddress;
        
      } else {
        options.error = 'invalid counter currency issuer: ' + params.counter.issuer;
        return;
      }
    }
    
  
    //Parse start and end times
    var time = tools.parseTimeRange(params.startTime, params.endTime, params.descending);
    
    if (time.error) {
      options.error = time.error;
      return;
    }
    
    if (!time.start || !time.end) {
      options.error = "startTime and endTime are required.";
    }
    
    options.startTime = time.start;
    options.endTime   = time.end;  
    
    
    //parse time increment and time multiple
    var results = tools.parseTimeIncrement(params.timeIncrement);  
    options.multiple = results.group_multiple;

        //set reduce option only if its false
    if (results.group_level)          options.view.group_level = results.group_level + 2;
    else if (params.reduce === false) options.view.reduce      = false;

    var limit  = params.limit  ? parseInt(params.limit, 10)  : 0,
      offset   = params.offset ? parseInt(params.offset, 10) : 0,
      maxLimit = 500,
      intervalCount;
      
    if (!limit || limit>maxLimit) limit = maxLimit;
      
    if (options.view.reduce===false) {
      if (limit  && !isNaN(limit))  options.view.limit = limit;
      if (offset && !isNaN(offset)) options.view.skip  = offset;
    } 
        
    if (results.group !== false) {
      intervalCount = tools.countIntervals(time.start, time.end, results.name);
      if (intervalCount>maxLimit) {
        return callback("Please specify a smaller time range or larger interval");
      }
    }
    
    // set startkey and endkey for couchdb query
    options.view.startkey   = [options.counter+":"+options.base].concat(options.startTime.toArray().slice(0,6));
    options.view.endkey     = [options.counter+":"+options.base].concat(options.endTime.toArray().slice(0,6));
    options.view.descending = params.descending || false; 
  }
} 


module.exports = offers;